feat(extras): added open_snippet_list

lyonelz96 opened the pull request:
Implements functionality discussed in #607

Here's a showcase of what it looks like:
![snippet_list](https://user-images.githubusercontent.com/43832900/200641196-bc536308-ccd9-4a85-9f64-bcdc7b501c15.gif)

A couple of thoughts:

The available and the get_context function are essentially the same as the ones in ```luasnip/init.lua``` the only difference is adding the docstring in get_context and using that new get_context in the available function. There might be a way to refactor something so that there is less code repetition if that's something you would be interested in.

I'm not great at regexes so I did the best I could with the folding stuff haha. I think it's fine right now but it can be a little finnicky, for example if you try to use telescopes buffer fuzzy finder in that window you can get something like this:
<img width="659" alt="image" src="https://user-images.githubusercontent.com/43832900/200642318-92c5885a-7b69-4216-96e0-c5bbaee5ca61.png">

At first I was just looking to open a new buffer and making that the current one, but making so that folding is available makes that hard since it's a per window and not a per buffer setting. So I went back to the vertical split implementation.

Let me know what you think, if everything seems good to you the only thing missing would be adding it to the documentation.

L3MON4D3 commented:
Nice!

> The available and the get_context function are essentially the same as the ones in luasnip/init.lua the only difference is adding the docstring in get_context and using that new get_context in the available function. There might be a way to refactor something so that there is less code repetition if that's something you would be interested in.

Oh, yes that would be great. Could extract the common part of getting available snippets, and then have the two functions collecting the information depend on that.

> I'm not great at regexes so I did the best I could with the folding stuff haha. I think it's fine right now but it can be a little finnicky, for example if you try to use telescopes buffer fuzzy finder in that window you can get something like this:

regex :vomiting_face: :P
Doesn't this example with the fuzzy finder always occur in nested folds? (independent of foldmethod? or are the folds not nested correctly?)
An alternative approach would be manual folding, then we'd need to take over part of `vim.inspect`'s job to reliably find the lines corresponding to each filetype

> At first I was just looking to open a new buffer and making that the current one, but making so that folding is available makes that hard since it's a per window and not a per buffer setting. So I went back to the vertical split implementation.

Good choice, very unfortunate that folding is per-window, that's a bit annoying (at least for us :D)

L3MON4D3 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,75 @@
+	local buf = vim.api.nvim_get_current_buf()
+	local win = vim.api.nvim_get_current_win()
+
+	vim.api.nvim_buf_set_name(buf, "Snippets")
```
I read somewhere that the buffer-name is tightly coupled to actual files on disk (found it, [here](https://github.com/neovim/neovim/issues/20531)) and common names are discouraged.
Using something URI-like makes sense, maybe `LuaSnip://Snippets` (maybe associate this with the buffer the snippets were queried for? `LuaSnip://Snippets (<bufnr>)` or something like that)

lyonelz96 replied:
Hmm interesting that ```vim.api.nvim_buf_set_option(buf, "buftype", "nofile")``` doesn't seem to do anything. I like that naming scheme better nonetheless.

Solved in https://github.com/L3MON4D3/LuaSnip/pull/652/commits/ab5e020f929e6e36c0f74452645455d80c307730. 
Let me know what you think of that name haha

L3MON4D3 replied:
xD
I think that's a bit long, maybe just `bufnr` or `buf` or `b`
Or make it configurable! That way, if someone wants to use something longer/shorter/different, it's just a lua-function away.

lyonelz96 replied:
😂 fair fair I've always liked long and descriptive names. Just to keep it simple we can go with the ```LuaSnip://Snippets buf <bufnr>``` so something like ```LuaSnip://Snippets buf 23``` for example. Just so people understand that the number corresponds to a buffer.

We could go the configurable route but I don't think it's worth adding the complexity even though it would be simple I think? But if we really want to go that route then we could default to the above name but if someone includes a name would you want to append to ```LuaSnip://Snippets``` or just replace it with whatever name the user gives it?

I'm thinking something simple like just adding an ```opts``` parameter to the ```open_snippet_list``` function. Opts being a table and one of the values being ```name```. So a user could do ```open_snippet_list({name = 'name here'})```

Hardest thing about software: Naming things 😂

L3MON4D3 replied:
> I'm thinking something simple like just adding an `opts` parameter to the `open_snippet_list` function. Opts being a table and one of the values being `name`. So a user could do `open_snippet_list({name = 'name here'})`

Hmm I'd just turn it into a lua-function which gets the bufnr as its lone argument, so
```lua
name = function(bufnr)
	return "LuaSnip://Snippets"
end

-- or
name = function(bufnr)
	return "LuaSnip://Snippets buf " .. bufnr
end
```
That keeps it very easy to extend (maybe insert buffer-name), and simple on our side.

> Hardest thing about software: Naming things :joy:

Oh true xD

L3MON4D3 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,75 @@
+	vim.diagnostic.disable(buf)
+
+	-- dump snippets
+	local replacements = vim.split(vim.inspect(snippets), "\n")
+	vim.api.nvim_buf_set_lines(buf, 0, 0, false, replacements)
+
+	-- enable local folding
+	vim.wo[win].foldmethod = "expr"
+	vim.wo[win].foldexpr = [[getline(v:lnum) =~ '= { {' ? 0 : 1]]
```
As I said in my answer, maybe manual folding will work better?
Additionally, some other/configurable way to pretty-print the snippets would be a cherry on top.

lyonelz96 replied:
So it turns out that setting the fold method to indent seems to work beautifully. Changes are in https://github.com/L3MON4D3/LuaSnip/pull/652/commits/fe3d9e29cce24af43a5aa0d3bc521c17ebf7b54f.

> Additionally, some other/configurable way to pretty-print the snippets would be a cherry on top.
I'm not entirely sure what you mean by this, would you mind expanding on that? You mean like a modifiable get_context function? Instead of returning a plain object we would return something else depending on parameters passed?

What is pretty-print to you?

L3MON4D3 replied:
> I'm not entirely sure what you mean by this, would you mind expanding on that?
> What is pretty-print to you?

Sure! So we have this big list of snippet-types -> filetypes -> these small context-tables, and currently use `vim.inspect` to turn that big table into a readable string, which is then put into the buffer.
`vim.inspect` includes `{`, `}`, `,`, <other characters> to turn its output into almost-valid lua, which we don't really need in our representation, we just want these context-tables grouped by snippet-type and filetype.
This could be achieved by writing a different function, which walks through the big table, and "pretty-prints" each table (so do something slightly different for each level)

For example, the docstrings don't really need quotes, as included by `vim.inspect`, and multiline-strings are tables, and thus look like `{"line1", "line2"}`, which could be improved.

I think it would be best to turn this into a configurable option, with `vim.inspect` as the default.
(we can include other pretty-printers in this module later on)

lyonelz96 replied:
> we just want these context-tables grouped by snippet-type and filetype

So ok let me make sure I'm understanding this correctly.
At the moment we have something like this:

```lua
{
    filetype = {
        snippet_context_table,
        snippet_context_table,
        etc...
    },
    filetype = {
        snippet_context_table,
        snippet_context_table,
        etc...
    },
}
```
and you're visualizing something more like this?

```lua
{
    filetype = {
        snippet_type = {
            snippet_context_table,
            etc...
        },
        snippet_type = {
            snippet_context_table,
            etc...
        },
    },
}
```

>  to turn its output into almost-valid lua, which we don't really need in our representation
> the docstrings don't really need quotes

So you're not visualizing this as a 'lua' type buffer? Just a text file?

> This could be achieved by writing a different function, which walks through the big table, and "pretty-prints" each table (so do something slightly different for each level)

So this function would be working on the context tables?

> multiline-strings are tables, and thus look like {"line1", "line2"}, which could be improved

Can you show me an example of where this happens?

>  this into a configurable option, with vim.inspect as the default.
(we can include other pretty-printers in this module later on)

So similar to the idea with configuring the buffer name, we could include a ```printer``` value in an opts table which is a function that grabs the big table that we get back from the ```available()``` function and turns it into the string that we want to output if I'm understanding you correctly.

I'm trying my best to understand what you're visualizing but I think I'm still confused I'm so sorry haha
Would it be possible for you to draw/sketch/pseucode what the output is supposed to look like in your eyes, I think that would help me make things clearer in my mind.

Sorry about that.

L3MON4D3 replied:
> > we just want these context-tables grouped by snippet-type and filetype
> 
> So ok let me make sure I'm understanding this correctly. At the moment we have something like this:
> 
> ```lua
> {
>     filetype = {
>         snippet_context_table,
>         snippet_context_table,
>         etc...
>     },
>     filetype = {
>         snippet_context_table,
>         snippet_context_table,
>         etc...
>     },
> }
> ```
> 
> and you're visualizing something more like this?
> 
> ```lua
> {
>     filetype = {
>         snippet_type = {
>             snippet_context_table,
>             etc...
>         },
>         snippet_type = {
>             snippet_context_table,
>             etc...
>         },
>     },
> }
> ```

Ah, disregard that, I thought the table was already grouped by snippet-type (regular vs autosnippet). This distinction might be nice to have, but it's not vital right now.

> > to turn its output into almost-valid lua, which we don't really need in our representation
> > the docstrings don't really need quotes
> 
> So you're not visualizing this as a 'lua' type buffer? Just a text file?

Yes, exactly. Similar to nvim-tree, but not of files and directories, but of snippets and file/snippet-types.

> > This could be achieved by writing a different function, which walks through the big table, and "pretty-prints" each table (so do something slightly different for each level)
> 
> So this function would be working on the context tables?

:+1:

> > multiline-strings are tables, and thus look like {"line1", "line2"}, which could be improved
> 
> Can you show me an example of where this happens?

Oh, `get_docstring` should return a `string[]`, weird that it doesn't show up that way..

> > this into a configurable option, with vim.inspect as the default.
> > (we can include other pretty-printers in this module later on)
> 
> So similar to the idea with configuring the buffer name, we could include a `printer` value in an opts table which is a function that grabs the big table that we get back from the `available()` function and turns it into the string that we want to output if I'm understanding you correctly.

Yes! default-value can be just `vim.inspect` since it has the correct arguments+output already.

> I'm trying my best to understand what you're visualizing but I think I'm still confused I'm so sorry haha Would it be possible for you to draw/sketch/pseucode what the output is supposed to look like in your eyes, I think that would help me make things clearer in my mind.

For example
```
lua
 ├─trigger: adsf
 │ docstring: adsf
 │ ...
 │ ...
 └─trigger: asdf2
   docstring: adsf2

python
 ├─trigger: adsf
 │ docstring: adsf
 │ ...
 │ ...
 └─trigger: asdf2
   docstring: adsf2
   ...
   ...
```
would look pretty cool

It's not necessary to include some advanced pretty-printer from the get-go, but implementing the configurability now would be great, then it's easy to improve later on.

lyonelz96 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,75 @@
+		description = snippet.dscr,
+		wordTrig = snippet.wordTrig and true or false,
+		regTrig = snippet.regTrig and true or false,
+		docstring = snippet.docstring,
```
Solved in https://github.com/L3MON4D3/LuaSnip/pull/652/commits/6c564cb1dde03ab1a34114f88b540947a06ded29

L3MON4D3 replied:
Ah, right, `:` not `.` :+1:

L3MON4D3 commented:
> Yeah! I don't know if you want to do this in this PR or in a separate one.

Doing that in this one should be fine, we introduce the potential for this abstraction here, I think that justifies it :D

> Seems that setting the foldmethod to indent fixed everything. I never use folding so I'm in uncharted waters here haha. If foldmethod set to indent is not enough I'll take a look into manual folding (I hope indent is enough lol)

Yeah, manual folding would probably be overcomplicating it :+1:

lyonelz96 commented:
> Doing that in this one should be fine, we introduce the potential for this abstraction here, I think that justifies it :D

Sounds good! I would like to get some of the stuff above resolved first since I feel stuff is piling up. But my initial thoughts on this was to 'inject' the get_context into the available function so something like:

```lua
function available(get_context)
    get_context = get_context or 'default function here'
end
````
Then we can use whatever get_context function we want while keeping the available function untouched.

L3MON4D3 commented:
> > Doing that in this one should be fine, we introduce the potential for this abstraction here, I think that justifies it :D
> 
> Sounds good! I would like to get some of the stuff above resolved first since I feel stuff is piling up. But my initial thoughts on this was to 'inject' the get_context into the available function so something like:
> 
> This would be in `luasnip/init.lua`
> 
> ```lua
> function available(get_context)
>     get_context = get_context or 'default function here'
> end
> ```
> 
> Then we can use whatever get_context function we want while keeping the available function untouched.

Oh yeah, nice!! :+1:

lyonelz96 commented:
(Sorry for the force pushes lol)

Ok so I bring some updates! Hopefully I'm heading in the right direction here

# Changes 

## luasnip/init.lua
In the available function I added a ```opts``` parameter which resolves to ```opts = opts or {get_context = get_context}``` that way we use the get_context that is defined in that file by default. Now that I'm thinking about it I think a better way would be to do
```lua
opts = opts or {}
opts.get_context = opts.get_context or get_context
```
Using the former would mean that if someone passes in an empty object for example ```available({})``` or any object ```available({blahblah = 'hihi'})``` opts wouldn't have any get_context entry. I'll update that.

But with that we have flexibility to use whatever get_context function we want

## extras/snippet_list.lua

I added an ```opts``` parameter to the open_snippet_list which resolves to 
```lua
opts = opts or {
    get_name = get_name,
    printer = printer,
    win_opts = win_opts(),
    buf_opts = buf_opts()
}
```

Same issue was above, it might be better to do something like

```lua
opts = opts or {}
opts.get_name = opts.get_name or get_name
etc...
```

### get_name function
It's as we discussed, asimple function that takes in a bufnr and returns the name for the buffer, you can pass in whatever function you would like that returns a string

### printer function
Takes in snippets and returns a string. At the moment we are only using ```vim.inpect```

### win_opts function
Returns a table of extra options to be set on the window. set_win_opts sets these options.

I use these 2 to set the folding at the moment

### buf_opts function
Returns a table of extra options to bet on the buffer. set_buf_opts sets these options.

I use these 2 to set the filetype to lua. (I know you want something like nvim tree but in the meantime I kept this here just to get syntax highlighting)

### make_scratch_buffer function
Just moved the setting options for the scratch buffer to this function to clean up the open_snippet_list function

So hopefully this is looking more like you were picturing in terms of configuration capability. 

# Discussion Points

## Pretty Print Stuff

> Similar to nvim-tree, but not of files and directories, but of snippets and file/snippet-types.

I think I get it now! Similar to what packer does when you sync/update plugins. Gotcha. That would definitely be awesome, atm it's a little over my head though. I'm not sure how nvim-tree or packer make stuff collapsible and get highlighting/icons for example. I would have to go dig in their repos and see what they are doing. 

In the meantime I left the folding method and the filetype to lua (For syntax highlighting purposes). With the configuration options we have now though it would be really easy to add whatever pretty-printer/buffer options/window options that you would like.

L3MON4D3 commented:
> (Sorry for the force pushes lol)

Oh, do not worry about that at all

> ## luasnip/init.lua
> In the available function I added a `opts` parameter which resolves to `opts = opts or {get_context = get_context}` that way we use the get_context that is defined in that file by default. Now that I'm thinking about it I think a better way would be to do
> 
> ```lua
> opts = opts or {}
> opts.get_context = opts.get_context or get_context
> ```
> 
> Using the former would mean that if someone passes in an empty object for example `available({})` or any object `available({blahblah = 'hihi'})` opts wouldn't have any get_context entry. I'll update that.
> 
> Edit: Done in [22e2370](https://github.com/L3MON4D3/LuaSnip/commit/22e2370952230c425bdb591d23a4c6afec82df05)

Nice! :+1:
WDYT about renaming it? I think `get_context` isn't 100% precise, but I also don't have a better idea (there we are again, naming things xD)

> ### get_name function
> It's as we discussed, asimple function that takes in a bufnr and returns the name for the buffer, you can pass in whatever function you would like that returns a string
> 
> ### printer function
> Takes in snippets and returns a string. At the moment we are only using `vim.inpect`
> 
> ### win_opts
> Table of extra options to be set on the window. set_win_opts sets these options.
> 
> I use these 2 to set the folding at the moment
> 
> ### buf_opts
> Table of extra options to set on the buffer. set_buf_opts sets these options.

Good idea to make buf/win-options configurable

> I use these 2 to set the filetype to lua. (I know you want something like nvim tree but in the meantime I kept this here just to get syntax highlighting)

Ahhh, yeah syntax highlighting makes sense.
If a different pretty-printer is used, one could write a custom syntax-file and use it by setting the filetype via buf-options, so that's pretty neat/complete.

> ### make_scratch_buffer function
> Just moved the setting options for the scratch buffer to this function to clean up the open_snippet_list function
> 
> So hopefully this is looking more like you were picturing in terms of configuration capability.
> 
> # Discussion Points
> ## Pretty Print Stuff
> > Similar to nvim-tree, but not of files and directories, but of snippets and file/snippet-types.
> 
> I think I get it now! Similar to what packer does when you sync/update plugins.

Yeeeeeeeeah right, that's even more common :+1:

> Gotcha. That would definitely be awesome, atm it's a little over my head though. I'm not sure how nvim-tree or packer make stuff collapsible and get highlighting/icons for example. I would have to go dig in their repos and see what they are doing.

Don't worry about that, this doesn't have to be perfect from the get-go. Updates to this can be done as non-breaking changes, so we're pretty flexible here imo.

> In the meantime I left the folding method and the filetype to lua (For syntax highlighting purposes). With the configuration options we have now though it would be really easy to add whatever pretty-printer/buffer options/window options that you would like.

Very clean! I'll take a look at the code now :D

L3MON4D3 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,92 @@
+	opts.buf_opts = opts.buf_opts or buf_opts
+
+	-- load snippets before changing windows/buffers
+	local available = require("luasnip").available
```
How about putting this outside the function, at the top of the file?

lyonelz96 replied:
Solved in https://github.com/L3MON4D3/LuaSnip/pull/652/commits/8cd564027f5fba9ab6d2af51da010dc051f71306

L3MON4D3 commented on lua/luasnip/init.lua:
```diff
@@ -61,19 +61,22 @@ local function get_context(snip)
-local function available()
+local function available(opts)
+	opts = opts or {}
+	opts.get_context = opts.get_context or get_context
```
`get_context` could be immediately stored in a local variable here, no need for the access through `opts`

L3MON4D3 replied:
Ohhh I see xD
`effective_get_context` would be nice (or change the current `get_context` to `default_get_context`, or something more descriptive (`docstring_context`?? since that's the difference to the original `get_context`))

lyonelz96 replied:
So as we discussed below, if you're ok with it, I can rename the original get_context to ```default_snip_info``` and the available function would look like

```lua
local function available(snip_info)
snip_info = default_snip_info or snip_info
end
```

L3MON4D3 replied:
Yeah that sounds good :+1:

lyonelz96 replied:
Solved in https://github.com/L3MON4D3/LuaSnip/pull/652/commits/868416192026216a556b62950c0463734b07f388

L3MON4D3 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,92 @@
+end
+
+local function get_name(bufnr)
+	return "LuaSnip://Snippets buf " .. bufnr
```
Would you be fine with making just `LuaSnip://Snippets` the default? The `LuaSnip://Snippets buf 0` looks a bit weird to me :sweat_smile:

lyonelz96 replied:
Solved in https://github.com/L3MON4D3/LuaSnip/pull/652/commits/3497f695a34301d2dac7c3b66235074361a5e39b

L3MON4D3 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,92 @@
+	return "LuaSnip://Snippets buf " .. bufnr
+end
+
+local function printer(snippets)
+	return vim.inspect(snippets)
+end
```
I don't think we have to wrap `vim.inspect` here, we could just `opts.printer = opts.printer or printer`.
(alternatively, `local printer = vim.inspect`)

lyonelz96 replied:
Solved in https://github.com/L3MON4D3/LuaSnip/pull/652/commits/5d2cd4505e27b19af24a6156a4daadcc28885c07

L3MON4D3 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,92 @@
+	local from_buf = vim.api.nvim_get_current_buf()
+
+	-- create and open buffer on right vertical split
+	vim.cmd("botright vnew")
```
WDYT about making this configurable? I feel like a function would be appropriate, just a vimscript-string might be too restrictive.
(not sure about name though, `open`? `open_new`?)

lyonelz96 replied:
Hmm, so if I'm understanding you correctly you want the user to have control over the window that is opened. So there is ```vim.api.nvim_open_win``` which has a wholleee lot of config options for the new window. We could default to ```vim.cmd("bothright vnew")``` and if the user wants to customize anything they could just pass whatever they would pass to ```vim.api.nvim_open_win``` to make the window however they want.

What do you think?

L3MON4D3 replied:
I think it might be more straightforward if we just say that "after `open`, the window (and buffer) the snippets should be displayed in, should both be current"

Another idea: we can pretty cleanly separate the steps we have to take here:
1. get info on available snippets (customize via `get_context`)
2. turn info into string (customize via `printer`)
3. display results of `printer` in some buffer (customize (currently) via `get_name`, `win_opts`, `buf_opts`).

We'd theoretically only need one option for the last one (to make it more abstract/capable), something like `display(printer_result, associated_bufnr)`, which could then open a split, a floating window, a new tab, add some weird highlighting to the buffer, whatever.

We could provide a default, which is basically just the code we have right now, and that itself (!!) could take the `get_name`, `win_opts`, `buf_opts`-parameters.
That should make this whole module very extensible, we could even implement a much more advanced `display` that collapses filetypes, has some interesting keybindings, ..., on this skeleton. (for such an advanced display we'd probably need some structured data from `printer`, so we should not restrain its result to `string`)

WDYT? Am I overcomplicating things? xD

lyonelz96 replied:
hahah I mean the original use case for myself was just to have a nicer way for me to search through the snippets to find the one I was looking for. Having all the customization is really awesome don't get me wrong but a bit overkill from my POV 😂. 

But from what I'm understanding we're not even changing the code that much just reorganizing a bit. You want to separate the ```open_snippet_list``` stuff into separate functions? Is that right?

I'm not sure what you want to do about the ```vim.cmd("botright vnew")```

> "after open, the window (and buffer) the snippets should be displayed in, should both be current"

That's what is happening already no? ```vim.cmd("botright vnew")``` opens a new vertical split with a buffer and makes both of those the current ones (Window and buffer).

You want to do something like this?

```lua
function open()
    vim.cmd("botright vnew")
    return vim.api.nvim_get_current_buf(), vim.api.nvim_get_current_win()
end
```

> Another idea: we can pretty cleanly separate the steps we have to take here:
    1. get info on available snippets (customize via get_context)
    2. turn info into string (customize via printer)
    3. display results of printer in some buffer (customize (currently) via get_name, win_opts, buf_opts).

```lua
function open_snippet_list(opts)
--- set opts (get_name, printer, win_opts, buf_opts)

-- Get info
local snippets = available(snip_info)

-- Stringify
local res  = printer(snippets)

-- Display
display(res, vim.api.nvim_get_current_buf())
-- In this function we would open the new window/buffer, set options, etc...
end
```

Something like that you mean? Or am I completely wrong?

L3MON4D3 replied:
> Having all the customization is really awesome don't get me wrong but a bit overkill from my POV :joy:

xD Yeah, I also felt like it might be too much, but it's a really nice design, and we have lots of flexibility.
If you'd like to just get this done, I totally understand, we don't need all that now, can easily extend

> That's what is happening already no?

Yes! I wanted to propose we use that as the definition for what `open` should do, but having `open` return the buffer+window as you suggest seems better (more standard, if you know what I mean).

> But from what I'm understanding we're not even changing the code that much just reorganizing a bit. You want to separate the `open_snippet_list` stuff into separate functions?

Yes, this would make it easier to replace just one of the parts (as a user, and for us, since we just have to provide a different set of functions, and those can then be plugged into `open_snippet_list` as `opts`)
My idea would be
```lua
function default_open_from_opts(opts) -- with opts get_name, win_opts, buf_opts
   return function(printer_result)
       <basically what we currently use for opening the buffer, setting the buffer-options, ...>
   end
end

function open_snippet_list(opts) ... end -- with opts get_info(snippet), printer(snip_info_by_ft_snippet_type), open(printer_result)
```

I think that might be better because it doesn't restrict how the snippets are displayed in any way, while also allowing the more straightforward customization we have currently (don't have to re-implement the whole default_open to just tweak a window-option)

lyonelz96 replied:
Check out https://github.com/L3MON4D3/LuaSnip/pull/652/commits/1bc38c70210e6be08a51ea850aab926736e8f590!

L3MON4D3 commented:
Okay, small things all, I like the general direction, nice work :+1:

lyonelz96 commented:
> Nice! 👍
WDYT about renaming it? I think get_context isn't 100% precise, but I also don't have a better idea (there we are again, naming things xD)

😂 Hmm just brainstorming here ```get_snip_context```, ```get_snip_tbl```, ```get_snip_info```, ```snip_context```, ```snip_tbl```, ```snip_info```, ```snip_inspect```, ```inspect_snip```

L3MON4D3 commented:
MHMM `snip_info` sounds good, not perfect though xD 
But I'd be fine with it

lyonelz96 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,92 @@
+
+	for opt, val in pairs(opts) do
+		vim.api.nvim_buf_set_option(buf, opt, val)
+	end
```
Solved in https://github.com/L3MON4D3/LuaSnip/pull/652/commits/05c5e31acc2d18cd726afad362c45d906a3f3e30

lyonelz96 commented:
Sorry for the inactivity. Have been a little busier. I made the modifications suggested! Hopefully it's closer to what we want :)

L3MON4D3 commented:
Oh, that's far from inactive, don't worry :D
From a cursory look, I like it, nice job!
I'll put some new comments in the code

L3MON4D3 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,88 @@
+	set_buf_opts(buf, opts)
+end
+
+local function default_open(printer_result)
```
I'd like to make it easier to adjust the "common" settings, like name of the buffer, or options (basically what we had earlier)
I think the best way to achieve this is to create a function which takes the options `win/buf_opts,get_name`, and returns `default_open` as a closure.
Basically
```lua
local function default_open(opts)
	-- maybe extend our options with opts.win/buf_opts instead? (overwrite on
	-- collision)
	local win_opts = opts.win_opts or win_opts
	local buf_opts = opts.buf_opts or buf_opts
	local get_name = opts.get_name or get_name

	return function(printer_result)
		-- use win_opts, buf_opts, get_name in here.
	end
```

the default (in `open_snippet_list`) could then just be `default_open()`.

lyonelz96 replied:
Oh I see! But wait wouldn't this be redundant since no one is calling default_open besides us no? Unless we pass ```opts``` to ```default_open``` too.

Cause otherwise it would look like this

```lua
local function open_snippet_list(opts)
	opts = opts or {}
	opts.snip_info = opts.snip_info or snip_info
	opts.printer = opts.printer or vim.inspect
	opts.open = opts.open or default_open()

	-- load snippets before changing windows/buffers
	local snippets = available(opts.snip_info)

	-- open snippets
	opts.open(opts.printer(snippets))
end
```

So options are never passed to ```default_open``` atm.

L3MON4D3 replied:
Ah, true, right now it wouldn't make a difference
We'd have to expose `default_open` :D
Then one can pass other defaults to it, and pass the generated function to `open_snippet_list`

lyonelz96 replied:
Exactly! Check https://github.com/L3MON4D3/LuaSnip/pull/652/commits/cbad3d030396415dc5e2d1d1fb79604937a02990

Now with this we can do something like

```lua
local snippet_list = require('luasnip.extras.snippet_list')
local open_snippet_list = snippet_list.open_snippet_list
local default_open = snippet_list.default_open

    open_snippet_list({
        open = default_open({
            buf_opts = {
                filetype = 'javascript',
            },
        }),
    })
```
As an example. Pretty neat. Of course the user can also define their own ```open``` function as well!

L3MON4D3 replied:
Niiice, that's great :ok_hand:

L3MON4D3 commented on lua/luasnip/init.lua:
```diff
@@ -510,21 +513,21 @@ local function unlink_current_if_deleted()
 	local snippet = node.parent.snippet
 	local ok, snip_begin_pos, snip_end_pos =
 		pcall(snippet.mark.pos_begin_end_raw, snippet.mark)
-	-- stylua: ignore
```
Oh, this section is ignored by stylua, could you manually revert it?
(same a few lines further down)

lyonelz96 replied:
Oh oops, I think I formatted it with sumneko instead of stylua. The dangers of having multiple formatters lol. Fixed in https://github.com/L3MON4D3/LuaSnip/pull/652/commits/2c1fda2ad554f6c674cd72df704fe58a3e7d237b hopefully.

L3MON4D3 replied:
Nice, that did it :+1:

L3MON4D3 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,96 @@
+
+return {
+	open_snippet_list = open_snippet_list,
+	default_open = default_open,
```
`default_open` -> `split` or `open_split`? That would be more descriptive, and fit better once we add more `open`-functions to this module

L3MON4D3 replied:
Maybe we should group the functions meant as `opts` into a subtable of `M`? (`snippet_list.options.open_split` looks good imo)

L3MON4D3 replied:
:ok_hand:

L3MON4D3 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,96 @@
+		local win = vim.api.nvim_get_current_win()
+
+		-- make scratch buffer
+		vim.api.nvim_buf_set_name(buf, opts.get_name())
```
Ah, I think the `buf` passed to `opts.get_name` was lost somewhere :sweat_smile:

lyonelz96 replied:
Oh yeah oops, the default get_name might not use it but it might be useful for users that write their own get_name function haha. Added it back in.

L3MON4D3 replied:
Perfect :ok_hand: :D

L3MON4D3 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,96 @@
+end
+
+return {
+	open_snippet_list = open_snippet_list,
```
Oh, and after seeing the example you typed out, I'd say just `open` sounds better, wdyt?

lyonelz96 replied:
I like it!

I changed ```opts.open``` to ```opts.open_split``` since it would look awkward to have the function be named open as well as the option. https://github.com/L3MON4D3/LuaSnip/pull/652/commits/4a4564ce8212ec3462b43be6e35661a586953e3c

L3MON4D3 replied:
Changing `open` is a good call, one issue I can see with `open_split` is that the function could also open a floating window, a new tab,...
How about `display` or `present` (present is a bit unusual, but it sounds cool imo :D) ?

lyonelz96 replied:
I actually like ```display``` if you're good with that!

L3MON4D3 replied:
Yep, go ahead👍

lyonelz96 replied:
Done! https://github.com/L3MON4D3/LuaSnip/pull/652/commits/ca044a48129c6539290ab82bec0c07d7c2441145

L3MON4D3 commented on lua/luasnip/extras/snippet_list.lua:
```diff
@@ -0,0 +1,96 @@
+
+return {
+	open = open,
+	options = { display = display },
```
Oh I think the option should be called `display`, but this name was fine, `open_split` (although open no longer fits exactly) is more descriptive than `display`.
Maybe `split_display` would be good, or something similar? (name should communicate that the function is meant for `display`, and that it opens a split)

lyonelz96 replied:
OH I gotcha, our default function opens a split so our function can be called ```split_display``` or ```display_split```?  but the name on opts can be the more general ```opts.display``` correct?

L3MON4D3 replied:
Yeah, exactly! :+1:

lyonelz96 replied:
Done. https://github.com/L3MON4D3/LuaSnip/pull/652/commits/35fd68e899d0a01f30ed16cffe8e2b73c3517cb2

L3MON4D3 commented:
Okay, I think this should be it, now.. Documentation :D
I think the existing entries for extras in `DOC.md` should be a good guideline, I'd recommend using `grip` to render the documentation locally (needs some setup, but is really nice for quickly iterating, as it renders markdown like github would)

lyonelz96 commented:
Added some docs for Snippet List! Let me know what you think. Once everything is ready to go I can squash the previous commits having to do with implementing snippet_list so it looks like `feat(extras): added snippet_list` and then we're good to go :D

L3MON4D3 commented on DOC.md:
```diff
@@ -1501,6 +1502,104 @@ callback will have access to the `POSTFIX_MATCH` field as well.
+Makes an `open` function available to use to open currently available snippets
+in a different buffer/window/tab.
+
+An `options` table with a `display` option (function) is also available in
+order to customize how you want the buffer/window/tab to look like. We'll take
+a deeper dive in the example below.
```
I'd prefer a more structured approach here, to improve readability.
So
* show the function-prototype
* list the arguments and possible options to those arguments

[example here](https://github.com/L3MON4D3/LuaSnip/blob/7bfae6fbedafa5572b5289b3dc2cba4cfde2039c/DOC.md?plain=1#L1219-L1240)

L3MON4D3 replied:
Oh, and explain `open` and the default-display separately, I think that will also make it more concise

lyonelz96 replied:
I addressed this in https://github.com/L3MON4D3/LuaSnip/pull/652/commits/34c03fa8c8e6836b13232111c7e33f98a015f881!

L3MON4D3 commented on DOC.md:
```diff
@@ -42,6 +42,7 @@ local fmt = require("luasnip.extras.fmt").fmt
 local fmta = require("luasnip.extras.fmt").fmta
 local conds = require("luasnip.extras.expand_conditions")
 local postfix = require("luasnip.extras.postfix").postfix
+local sl = require("luasnip.extras.snippet_list")
```
Oh, this list is mainly meant for shorthands used when defining snippets, and I don't think `sl` is useful there :D
Maybe introduce this shorthand in the `snippet_list`-section itself?

lyonelz96 replied:
Moved the shorthand to the snippet list section!

lyonelz96 commented on DOC.md:
```diff
@@ -1501,6 +1502,104 @@ callback will have access to the `POSTFIX_MATCH` field as well.
+
+-- opts.win_opts -> table which entries have a {window_option -> value} form
+
+-- opts.get_name -> function that is passed the buffer handle and returns a
```
Check out the revamp at https://github.com/L3MON4D3/LuaSnip/pull/652/commits/34c03fa8c8e6836b13232111c7e33f98a015f881

Much much cleaner imo

L3MON4D3 commented on doc/luasnip.txt:
```diff
@@ -1453,6 +1454,152 @@ callback will have access to the `POSTFIX_MATCH` field as well.
+<p class="caption">display</p>
+</div>
+
+An `options` table with a `display` option (function) is also available in
+order to customize how you want the buffer/window/tab to look like. We’ll
+take a deeper dive in the example below.
```
Much better!!

I think this section is a bit unclear though, maybe first introduce the `options`-table (+motivation), and then, in a list, mention the different functions in that table. (maybe also motivate why `display` is included from the get-go, should be easy with the big block of code before :D) (ouh, maybe construct the previous custom display in such a way that it can also be achieved via arguments to `options.display`, then its purpose should be immediately clear)

lyonelz96 replied:
Is this https://github.com/L3MON4D3/LuaSnip/pull/652/commits/82680e6306e16e63cba3af4225da2a3a5537e9a5 more what you had in mind?

L3MON4D3 replied:
Yeah, I like it👍
I think as-is, the text for that bullet point is not in the list, some indent should fix that (might be wrong though, on mobile atm)

L3MON4D3 commented on doc/luasnip.txt:
```diff
@@ -1453,6 +1454,152 @@ callback will have access to the `POSTFIX_MATCH` field as well.
+- `opts`: optional arguments:
+    - `snip_info`: `snip_info(snippet) -> table representation of snippet`
+    - `printer`: `printer(snippets:table) -> string representation of snippets`
+    - `display`: `display(snippets:string)`
```
I think we should ignore that `printer` returns a string by default, and allow it to return anything. (we might want to have structured data in `display` when building some kind of tree, and then `printer` would not be necessary (e.g., it would be the identity-function))

lyonelz96 replied:
Hmm this is interesting because by default we're kinda counting on `printer` returning a string. So the only scenario where printer could NOT return a string would be if the user implemented their own `display` as well.

So maybe I should add another section explaining that by default `display` is called with the result from `printer` and if you want to return a table that has a certain structure from `printer` or hell return anything you'd have to implement your own `display` that parses that input OR like you said you could just implement your own display and do whatever you want in there without even using printer.

Hmm what do you think?

L3MON4D3 replied:
I've been thinking along the same line(is there a reason for separating display and printer), but I think it makes sense to separate them since the most common return type of printer is a string, and then all displays accepting a string are compatible (otherwise, there'd be some duplicate code in the displays)
I think going into more detail on this would be great👍👍

lyonelz96 replied:
Is this https://github.com/L3MON4D3/LuaSnip/pull/652/commits/74f508ff91e54f530d7b795807587c366c4ab0e4 sufficient you think?

L3MON4D3 replied:
Nice! Maybe parse -> stringify, in the last line (at least I think that's more like what you meant, am I right?)

lyonelz96 replied:
Yeah that's better! I fixed it. If that's everything, let me know and I'll squash the commits and separate it into just two different commits (Code Implementation and Documentation)

L3MON4D3 replied:
Okay awesome, this is it :partying_face::partying_face:
Two works :ok_hand:

L3MON4D3 commented:
Hey, very sorry for not getting back earlier, I must've missed your previous messages somehow, and didn't think to check on this PR until now😅

lyonelz96 commented:
> Hey, very sorry for not getting back earlier, I must've missed your previous messages somehow, and didn't think to check on this PR until now😅

No worries! Thought you might be busy haha. I'll be travelling tomorrow but I'll take a look over the next few days see if we can finish this off before christmas or new years. We're almost there!

L3MON4D3 commented:
Nice, sounds good :+1: :D

lyonelz96 commented:
Squashing of commits is done!

L3MON4D3 commented:
Okay I'm so sorry, I missed two minor issues:

 * pictures are embedded in vim-help, which is pretty ugly (fix: surround with ignore-tags, look ar source of gifs to see what I mean)
 * no doc for new capabilities of `ls.available` (I mean it's not that useful in itself, but there's no reason to not document it)

I'm taking another 10000%-close look to make sure those are the last roadblocks, would you clear them? :D

L3MON4D3 commented:
Okay, 10000%-close look done, everything else looks great :)

lyonelz96 commented:
> Okay I'm so sorry, I missed two minor issues:
> 
>     * pictures are embedded in vim-help, which is pretty ugly (fix: surround with ignore-tags, look ar source of gifs to see what I mean)
> 
>     * no doc for new capabilities of `ls.available` (I mean it's not that useful in itself, but there's no reason to not document it)
> 
> 
> I'm taking another 10000%-close look to make sure those are the last roadblocks, would you clear them? :D

Done!
Hopefully I didn't miss any pictures lol.
Checkout the updated description for the available function, I don't know if that's clear enough?

L3MON4D3 commented on DOC.md:
```diff
@@ -2809,8 +2809,14 @@ the lazy_load.
+- `available(snip_info)`: return a table of all snippets defined for the
+  current filetypes(s) (`{ft1={snip1, snip2}, ft2={snip3, snip4}}`).
+  The structure of the snippet is defined by `snip_info` which is a function
+  (`snip_info(snip)`) that takes in a snippet (`snip`) and uses it to define
+  the desired table.
```
Maybe `... , finds the desired information on it, and returns it`

lyonelz96 replied:
Like this you mean? ```(`snip_info(snip)`) that takes in a snippet (`snip`), finds the desired
  information on it, and returns it.``` ?

L3MON4D3 replied:
Yeah, something like it👍
I'd like this passage to have more context, `uses it to define the desired table` sounds a bit too general imo

lyonelz96 replied:
Done! https://github.com/L3MON4D3/LuaSnip/pull/652/commits/5eace87c76cd733bc8f0be71e17098a88e408d11

L3MON4D3 commented:
Pictures look good, doc too except that small nit, thanks for taking care of it :)

L3MON4D3 commented:
Haa okay, that's it, perfect, I've got nothing more to ask, you're free xD
One more squash, rebase if you wanna, then I'll merge and we've done it :)))))

lyonelz96 commented:
> Haa okay, that's it, perfect, I've got nothing more to ask, you're free xD One more squash, rebase if you wanna, then I'll merge and we've done it :)))))

Squash done!

![Dobby](https://media.giphy.com/media/wUrc9zZpRhRrW/giphy.gif)

L3MON4D3 commented:
Hah, nice xD

Okay, thank you sooo much for sticking through this entire process. I feel like I've asked a bit too much, and those small redesigns we did halfway through could've been separate PRs, but I'm very happy with this being the initial result because it's very flexible and we should be able to extend it without any breaking changes.
If you have any other ideas, feel free to open an issue to discuss it :)

lyonelz96 commented:
> Hah, nice xD
> 
> Okay, thank you sooo much for sticking through this entire process. I feel like I've asked a bit too much, and those small redesigns we did halfway through could've been separate PRs, but I'm very happy with this being the initial result because it's very flexible and we should be able to extend it without any breaking changes. If you have any other ideas, feel free to open an issue to discuss it :)

No problem! Glad I could contribute a small part to this plugin. Thank you for your work!
